processor/src/chiplets/aux_trace/bus/kernel.rs

```diff
@@ -27,33 +61,43 @@ pub(super) fn build_kernel_chiplet_responses<E>(
 where
     E: FieldElement<BaseField = Felt>,
 {
-    // The caller ensures this row is a kernel ROM row, we just check this is a bus response row.
-    if main_trace.chiplet_kernel_is_bus_response(row) {
-        let message = {
-            let root0 = main_trace.chiplet_kernel_root_0(row);
-            let root1 = main_trace.chiplet_kernel_root_1(row);
-            let root2 = main_trace.chiplet_kernel_root_2(row);
-            let root3 = main_trace.chiplet_kernel_root_3(row);
-
-            KernelRomMessage {
-                kernel_proc_digest: [root0, root1, root2, root3],
-            }
-        };
+    let root0 = main_trace.chiplet_kernel_root_0(row);
+    let root1 = main_trace.chiplet_kernel_root_1(row);
+    let root2 = main_trace.chiplet_kernel_root_2(row);
+    let root3 = main_trace.chiplet_kernel_root_3(row);
 
+    // The caller ensures this row is a kernel ROM row, so we just need to check if this is
+    // the first row for a unique proc hash.
+    if main_trace.chiplet_kernel_is_first_hash_row(row) {
+        // Respond to the requests performed by the verifier when they initialize the bus
+        // column with the unique proc hashes.
+        let message = KernelRomInitMessage {
+            kernel_proc_digest: [root0, root1, root2, root3],
+        };
         let value = message.value(alphas);
 
         #[cfg(any(test, feature = "bus-debugger"))]
-        _debugger.add_response(alloc::boxed::Box::new(message), alphas);
+        _debugger.add_response(std::boxed::Box::new(message), alphas);
 
         value
     } else {
-        E::ONE
+        // Respond to decoder messages.
+        let message = KernelRomMessage {
+            kernel_proc_digest: [root0, root1, root2, root3],
+        };
+        let value = message.value(alphas);
+
+        #[cfg(any(test, feature = "bus-debugger"))]
+        _debugger.add_response(std::boxed::Box::new(message), alphas);
+        value
     }
 }
 
-// MESSAGE
+// MESSAGES
 // ===============================================================================================
 
+/// A message between the decoder and the kernel ROM to ensure a SYSCALL can only call procedures
+/// provided through public inputs.
```

Comment:
nit: can only call procedures in the kernel as specified through public inputs ( or something similar)